src/server/serverinventorymgr.cpp

```diff
@@ -0,0 +1,198 @@
+		delete detached_inventory.second;
+	}
+}
+
+Inventory *ServerInventoryMgr::getInventory(const InventoryLocation &loc)
+{
+	switch (loc.type) {
+	case InventoryLocation::UNDEFINED:
+	case InventoryLocation::CURRENT_PLAYER:
+		break;
+	case InventoryLocation::PLAYER: {
+		RemotePlayer *player = m_env->getPlayer(loc.name.c_str());
+		if (!player)
+			return NULL;
+		PlayerSAO *playersao = player->getPlayerSAO();
+		if (!playersao)
+			return NULL;
+		return playersao->getInventory();
+	} break;
+	case InventoryLocation::NODEMETA: {
+		NodeMetadata *meta = m_env->getMap().getNodeMetadata(loc.p);
+		if (!meta)
+			return NULL;
+		return meta->getInventory();
+	} break;
+	case InventoryLocation::DETACHED: {
+		if (m_detached_inventories.count(loc.name) == 0)
+			return NULL;
+		return m_detached_inventories[loc.name];
+	} break;
+	default:
+		sanity_check(false); // abort
+		break;
+	}
+	return NULL;
+}
+
+void ServerInventoryMgr::setInventoryModified(const InventoryLocation &loc)
+{
+	switch (loc.type) {
+	case InventoryLocation::UNDEFINED:
+		break;
+	case InventoryLocation::PLAYER: {
+
+		RemotePlayer *player = m_env->getPlayer(loc.name.c_str());
+
+		if (!player)
+			return;
+
+		player->setModified(true);
+		player->inventory.setModified(true);
+		// Updates are sent in ServerEnvironment::step()
+	} break;
+	case InventoryLocation::NODEMETA: {
+		MapEditEvent event;
+		event.type = MEET_BLOCK_NODE_METADATA_CHANGED;
+		event.p = loc.p;
+		m_env->getMap().dispatchEvent(event);
+	} break;
+	case InventoryLocation::DETACHED: {
+		// Updates are sent in ServerEnvironment::step()
+	} break;
+	default:
+		sanity_check(false); // abort
+		break;
+	}
+}
+
+Inventory *ServerInventoryMgr::createDetachedInventory(
+		const std::string &name, IItemDefManager *idef, const std::string &player)
+{
+	if (m_detached_inventories.count(name) > 0) {
+		infostream << "Server clearing detached inventory \"" << name << "\""
+			   << std::endl;
+		delete m_detached_inventories[name];
+	} else {
+		infostream << "Server creating detached inventory \"" << name << "\""
+			   << std::endl;
+	}
+
+	Inventory *inv = new Inventory(idef);
+	sanity_check(inv);
+	m_detached_inventories[name] = inv;
+	if (!player.empty()) {
+		m_detached_inventories_player[name] = player;
+
+		// if player is connected, send him the inventory
+		if (RemotePlayer *p = m_env->getPlayer(name.c_str()))
+			m_env->getGameDef()->sendDetachedInventory(
+					inv, name, p->getPeerId());
+	} else {
+		// Inventory is for everybody, broadcast
+		m_env->getGameDef()->sendDetachedInventory(inv, name, PEER_ID_INEXISTENT);
+	}
+
+	return inv;
+}
+
+bool ServerInventoryMgr::removeDetachedInventory(const std::string &name)
+{
+	const auto &inv_it = m_detached_inventories.find(name);
+	if (inv_it == m_detached_inventories.end())
+		return false;
+
+	delete inv_it->second;
+	m_detached_inventories.erase(inv_it);
+
+	if (!m_env) // Mods are not done loading
```

Comment:
Since `m_env` is set in the class constructor, this check never does anything.

Reply:
yep exact, it was just a code move, maybe a historical ugly hack :)